rbd/mirror_nautilus.go

```diff
@@ -0,0 +1,379 @@
+		State:       MirrorImageState(status.info.state),
+		StatusState: MirrorImageStatusState(status.state),
+		GlobalID:    C.GoString(status.info.global_id),
+		IsPrimary:   bool(status.info.primary),
+		IsUp:        bool(status.up),
+	}, nil
+}
+
+// MirrorImageList returns a MirrorImageInfo for each mirrored image.
+//
+// Note: this can only be used if go-ceph is compiled with the `nautilus` build
+// tag.
+//
+// Implements:
+//  int rbd_mirror_image_status_list(rados_ioctx_t io_ctx,
+//  				      									 const char *start_id, size_t max,
+//  				      									 char **image_ids,
+//  				      									 rbd_mirror_image_status_t *images,
+//  				      									 size_t *len)
+func MirrorImageList(ioctx *rados.IOContext) ([]*MirrorImageInfo, error) {
+	imageInfos := make([]*MirrorImageInfo, 0)
+	const cMaxIter C.size_t = 100
+	var startID string
+
+	for {
+		// We need to wrap all the actions within the for loop in a function
+		// in order to ensure that we correctly reclaim all allocated memory
+		// from C at the end of every iteration.
+		ret, done := iterateImageList(ioctx, &imageInfos, &startID, cMaxIter)
+		if ret != 0 {
+			return imageInfos, getError(ret)
+		}
+
+		if done {
+			break
+		}
+	}
+	return imageInfos, nil
+}
+
+func iterateImageList(ioctx *rados.IOContext, imageInfos *[]*MirrorImageInfo, startID *string, cMaxIter C.size_t) (C.int, bool) {
+	cImageIDs := make([]*C.char, cMaxIter)
+	cImageStatus := make([]C.rbd_mirror_image_status_t, cMaxIter)
+	done := false
+
+	var cLen C.size_t
+	ret := C.rbd_mirror_image_status_list(cephIoctx(ioctx), C.CString(*startID),
+		cMaxIter, &cImageIDs[0], &cImageStatus[0], &cLen)
+	if ret != 0 {
+		return ret, done
+	}
+
+	// If the list length is 0 or less than the max size
+	// specified we know we are on the last page of the list,
+	// and we don't need to continue iterating.
+	if cLen < cMaxIter {
+		done = true
+	}
+
+	if cLen == 0 {
+		return C.int(0), done
+	}
+
+	defer func() {
+		C.rbd_mirror_image_status_list_cleanup(&cImageIDs[0], &cImageStatus[0], cLen)
+	}()
+
+	for i := 0; i < int(cLen); i++ {
+		mi := &MirrorImageInfo{
+			Name:        C.GoString(cImageStatus[i].name),
+			Description: C.GoString(cImageStatus[i].description),
+			State:       MirrorImageState(cImageStatus[i].info.state),
+			StatusState: MirrorImageStatusState(cImageStatus[i].state),
+			GlobalID:    C.GoString(cImageStatus[i].info.global_id),
+			IsPrimary:   bool(cImageStatus[i].info.primary),
+			IsUp:        bool(cImageStatus[i].up),
+		}
+
+		*imageInfos = append(*imageInfos, mi)
+	}
+
+	*startID = C.GoString(cImageIDs[cLen-1])
+	return C.int(0), done
+}
+
+// MirrorEnable will enable mirroring for an image.
+//
+// Note: this can only be used if go-ceph is compiled with the `nautilus` build
+// tag.
+//
+// Implements:
+//  int rbd_mirror_image_enable(rbd_image_t image)
+func (image *Image) MirrorEnable() error {
+	ret := C.rbd_mirror_image_enable(C.rbd_image_t(image.image))
```

Comment:
`image.image` is of type `C.rbd_image_t` already. Also check if you need `validate` code from comment above.

Reply:
Done!